### DISCUSSION TITLE
Convert a group of .psd or similar format, label each image (with a slight drop shadow) and write out to multiple jpg images in different resolutions.

### DISCUSSION THREAD

**Discussion by ed42311** [ImageMagick/ImageMagick#2154] [Help]
Hello imagemagick community, I've been working on a script for a little while now (not gonna say how long 😆 ) that in simple terms is meant to do what is described above.  It may be useful to note that this script is running in a node instance as it is eventually used in a deployment script through GH Actions.

My current assumption is that the IM part of the operation is independent of the environment in which it is being run, but yaknow, could be wrong about that...  Only heard about imagemagick when I started working on this project so I'm beginning to understand some things, but the vast majority of it is a mystery.

Output of `convert --version`

```bash
Version: ImageMagick 7.0.10-13 Q16 x86_64 2020-05-17 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(3.1)
Delegates (built-in): bzlib freetype gslib heic jng jp2 jpeg lcms ltdl lzma openexr png ps tiff webp xml zlib
```

This is my current iteration:

> The only items that vary when run are __derivedSize__ which I gather from the original image and pass to the next process and __storedGravity__ which is coming from a json object and is tied as data to the original image.  The __derivedSize__ option was to solve a problem of the label being set by some default and coming out at varying heights.

```bash
convert -background none -fill color -density 90 -size x<derivedSize> \
-font "Helvetica-Bold" label:<storedLabel> \
\( +clone -background black -shadow 90x3+10+10 \) +swap \
-background none -layers merge \( <pathtofile> -flatten -strip \) +swap \
-gravity <storedGravity> -geometry +30+30 -composite +repage PNG00:- \
- -write mpr:image +delete \( mpr:image -resize 240 -write <newfile>@1x.jpg \) \
\( mpr:image -resize 480 -quality 92% -write <newfile>@2x.jpg \) \
\( mpr:image -resize 960 -quality 82% -write <newfile>@4x.jpg )
```

## Constraints

1. Label should rest in a designated corner by gravity, padded in from the corner a consistent amount across images
2. Label should be sharp edged and pop out of the frame a bit, currently achieved with the shadow, this is to maintain legibility and to make sure that, for instance white lettering does not fade into a similar color in the image frame
3. Font size should hover around 14-18px at 240 width, in pointsize the ideal seems to be somewhere around 16pt

## Issues I'm encountering:

1. The label seems to have a slight blurriness which I'm unable to get rid of.  Thsi is maybe mostly do to the conversion to jpeg?  Not sure
2. The position seems to shift based on the image, which is a little confusing, maybe I'm getting the geometry wrong?

## Things I've tried by issue

1. 
    1. _tweak the shadow_ - this seems to help a little bit, but has a hard limit, also it appears to be most effective at larger font size and I'd like to try and keep the relative font size around 
    2. _remove the quality modifiers on write_ - this does seem to improve things slightly, which is expected... 😸 
    3. _remove the mpr:image action and write each image out individually_ -  this isn't great because it just takes a bunch more node processes, I kind of liked that you could pass the image off this way, but maybe there is a better way that retains fidelity
2.
    1. _mostly just messing around with the geometry settings_ 

## Resources referenced

1. https://imagemagick.org/Usage/text/
2. https://imagemagick.org/discourse-server/viewtopic.php?t=25073 - which makes me think maybe I don't need mpr
3. http://www.imagemagick.org/discourse-server/viewtopic.php?t=35202

## Possible alternative paths

1. Annotate ? - does this provide any better fidelity?
2. clone instead of mpr
3. maybe the underlying problem is simple the lossy format at the end of the line?

**Comment by fmw42** [comment]
What is your ImageMagick version? (convert -version). Can you provide an exact command line with your arguments and input and output so that we can try to duplicate your result

**Reply by ed42311** [reply]
```bash
Version: ImageMagick 7.0.10-13 Q16 x86_64 2020-05-17 https://imagemagick.org
Copyright: © 1999-2020 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(3.1)
Delegates (built-in): bzlib freetype gslib heic jng jp2 jpeg lcms ltdl lzma openexr png ps tiff webp xml zlib
```

interestingly the `-` didn't work so I switched to clone

```bash
convert -background none -fill white -density 90 -size x250 -font "Helvetica-Bold" label:Hello \( +clone -background black -shadow 90x3+10+10 \) +swap -background none -layers merge \( beautiful_grass_psd.psd -flatten -strip \) +swap -gravity SouthWest -geometry +30+30 -composite +repage \( -clone 0 -resize 240 -write newfile@1.jpg \) \( -clone 0 -resize 480 -write newfile@2.jpg \) \( -clone 0 -resize 960 -write newfile@4.jpg \) null:
```

![newfile@1](https://user-images.githubusercontent.com/14878694/84806528-18eead00-afc3-11ea-9bc9-5f8b0b15ec3e.jpg)
![newfile@2](https://user-images.githubusercontent.com/14878694/84806544-1c823400-afc3-11ea-90eb-e114676a8d6b.jpg)
![newfile@4](https://user-images.githubusercontent.com/14878694/84806552-1ee48e00-afc3-11ea-891e-a48510b50cff.jpg)

I don't seem to be able to upload the psd but these are the results ^

**Reply by ed42311** [reply]
Any sort of natury psd will do...  you can see there is just a bit of a blur on the edges of the text

**Reply by ed42311** [reply]
and then the padding goes all screwy when I switch gravity, although that might be a misunderstanding on my part as to how `-geometry` affects `-gravity`

**Reply by fmw42** [reply]
You can zip the PSD file and upload the zip file. With ImageMagick 7, use magick, not convert. I do not see what is the issue with gravity in your posted images. Please clarify. Does your PSD file have more than one layer? Best if you can post the PSD file. The -geometry sets the offset from the specified -gravity. Positive values in -geometry move the placement of the upper left corner of the text image away from corners and towards the center, unless -gravity is center. Then the values move from the center.

**Reply by ed42311** [reply]
Don't think I can share the actual PSD, but yes multilayer, I'll see what I can share... 

With that one specific gravity it doesn't seem to be a problem - the full node script runs through 150 images and uses different gravities, but does not change geometry. You said upper left corner, towards the center... always upper left ?  I was thinking about it more like a bounding box and padding.  Maybe that model is incorrect. 

Then the more salient issue is probably the fuzziness on the text, I thought increasing `-density` would solve it, but that does not seem to be the case.

**Reply by dlemstra** [reply]
You are "upsizing" the first image? The first image is `x250` and then resized to `960`. What happens when you start with `x960`?

**Reply by ed42311** [reply]
yes, ok, but then the text is crazy big, do I need to make a place each textual element at the size of the final image?
![newfile@1](https://user-images.githubusercontent.com/14878694/84925143-bb239900-b086-11ea-97e5-2e4b3c51aad0.jpg)

**Reply by ed42311** [reply]
Maybe I need to move some of these actions around, saw an example using `-compose over`

or maybe make a new text block for each written image, whew that seems like a lot of work... 

I am running this in a node instance so I could break this into steps and make label pngs for each size, write them out to a temp folder along with the converted PNG of the PSD then gather them back up and write them out... ugh, not very performant but it's a possibility, I'll take a look tonight.

**Comment by snibgo** [comment]
The text is "-size x250", and one final image after compositing the text over beautiful_grass_psd.psd is "-resize 960", but we can't tell whether this enlarges or shrinks without knowing the size of beautiful_grass_psd.psd. If this enlarges, you will get fuzziness.

I can't see any fuzziness in the text. It is anti-aliased, of course. If you don't want that, insert "+antialias" before "label:Hello". Shrinking will introduce a form of anti-aliasing. You could sharpen the image after shrinking, if you want. Or use a shrinking method that includes a sharpening. See [Resampling with halo minimization](http://im.snibgo.com/resamphm.htm).
